Reload PropertiesEndpointGroup if the watched properties file is updated

jrhee17 opened the pull request:
Purpose:
- Give option to have PropertiesEndpointGroup reload automatically if the watched properties file is updated

Modification:
- Add a constructor in `PropertiesEndpointGroup` which receives a `filePath`. Allow `endpoints` dynamic update by inheriting `DynamicEndpointGroup`.
- Add a `FileWatcherRegistry` class which keeps track of each file watch event.
- Add a `RestartableThread` class which runs a new `Runnable` every time it is restarted. 
- Add a `FileWatcherRunnable` class which is responsible for detecting file changes in background and calling the appropriate callback method.

Result:
- When a `PropertiesEndpointGroup` is initialized specifying the properties path, the `endpointGroup` will reload automatically on changes to the properties file.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/line/armeria?pullRequest=1787) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/line/armeria?pullRequest=1787) before we can accept your contribution.<br/><hr/>**John** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/line/armeria?pullRequest=1787) it.</sub>

hyangtack commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,173 @@
+     */
+    void deregister(@Nonnull URL resourceUrl) {
+        final PropertiesFileWatcherContext context = ctxRegistry.remove(resourceUrl.getFile());
+        context.key.cancel();
```
Don't we need to `null` check to avoid NPE? (It seems possible to be called more than once )

anuraaga commented:
@imasahiro I remember you were working on something related to file watching - is there any context we've forgotten about since then? :)

anuraaga commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,173 @@
+                        final String watchedPathFile = watchedPath.toFile().getAbsolutePath();
+
+                        if (event.kind().equals(ENTRY_MODIFY) || event.kind().equals(ENTRY_CREATE)) {
+                            if (ctxRegistry.containsKey(watchedPathFile)) {
```
There are certain environments where this simple check doesn't actually work from what I've seen. The file might be a symlink, or the directory, not the file might change. I've had better luck iterating, and comparing the real path using `startsWith`

https://github.com/curioswitch/curiostack/blob/master/common/server/framework/src/main/java/org/curioswitch/common/server/framework/files/FileWatcher.java#L140

anuraaga commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,173 @@
+
+    private final Map<String, PropertiesFileWatcherContext> ctxRegistry = new ConcurrentHashMap<>();
+    private final WatchService watchService;
+    private final ExecutorService eventLoop =
```
We use `EventLoop` throughout the codebase with a very specific meaning (Netty event loop), it would be better to call this `executor` to prevent confusion.

anuraaga commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,223 @@
+                while ((key = watchService.take()) != null) {
+                    for (WatchEvent<?> event : key.pollEvents()) {
+                        @SuppressWarnings("unchecked")
+                        final Path watchedPath = getRealPath(
```
Think you should catch and log the exception here and `continue` - it's possible `getRealPath` will throw if the file was modified then immediately deleted. Fine to skip such an event.

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroup.java:
```diff
@@ -85,13 +93,42 @@ public static PropertiesEndpointGroup of(ClassLoader classLoader, String resourc
+     * Creates a new {@link EndpointGroup} instance that loads the host names (or IP address) and the port
+     * numbers of the {@link Endpoint} from the {@code resourceName} resource file. The resource file must
+     * contain at least one property whose name starts with {@code endpointKeyPrefix}:. Reloading property files
+     * for changes is {@code false} by default.
```
`Reloading property files ... by default.` could be removed? If you want to emphasize that the other factory methods do not reload, you could write something there, not here.

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroup.java:
```diff
@@ -36,9 +40,11 @@
 
-    // TODO(ide) Reload the endpoint list if the file is updated.
+    @Nullable
+    private Runnable closeCallback;
```
- This field has to be moved after the static methods.
- This field could be `final` if it is set with `null` when unused.

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroup.java:
```diff
@@ -181,14 +247,11 @@ private static void validateDefaultPort(int defaultPort) {
-    public List<Endpoint> endpoints() {
-        return endpoints;
+    public void close() {
+        super.close();
```
Maybe we should wrap with `try-finally` and call `closeCallback` in the `finally` block?

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,215 @@
+    private final Map<Path, PropertiesFileWatcherContext> ctxRegistry = new ConcurrentHashMap<>();
+    private final WatchService watchService;
+    private final ExecutorService executor =
+            Executors.newSingleThreadExecutor(new ThreadFactoryBuilder().setDaemon(true).build());
```
- Could you set the thread name?
- Could we make the executor is started lazily and shut down when there are no watches anymore, so that this class is more friendly to multi-classloader environment? For example, someone might be using Armeria as a part of a Java EE webapp and want Armeria to clean up everything correctly when the web application is reloaded.

jrhee17 replied:
I was kind of thinking as long as a submitted `thread` (or `future`) isn't running, there should be minimal overhead from just instantiating `ExecutorService`. (The `future` is currently stopped when no paths are registered). Let me know if I am thinking about this wrong 🤔

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,215 @@
+    synchronized void register(Path filePath, Runnable reloader) {
+        final Path realFilePath = getRealPath(filePath);
+        checkArgument(!ctxRegistry.containsKey(realFilePath),
+                      "file is already watched: %s", realFilePath);
```
What happens if a user creates a two `PropertiesEndpointGroup` instances with the same paths? Although not very efficient, we would still allow such a case.

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,215 @@
+                                }
+                            }
+                        } else if (event.kind().equals(OVERFLOW)) {
+                            logger.info("failed to reload file: {}", realPath);
```
I'm not sure which is correct, but maybe we can try reloading in this case as well?

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,215 @@
+     */
+    PropertiesFileWatcherRegistry() {
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
```
If I understood the NIO.2 API correctly, you have to use the `FileSystem` retrieved from `Path`:

```java
final Path path = Paths.get(uri);
watchService = path.getFileSystem().newWatchService();
```

This means, we may have more than one `WatchService` implementations depending on how a user configured the JRE, i.e. this class needs some refactoring to have different `PropertiesFileWatchRegistry`s for different `FileSystem`s.

jrhee17 replied:
I'm still thinking about use-cases for this & how I can do this efficiently 🤔. Just leaving a comment to let you know I'm not disregarding this comment:)

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroupTest.java:
```diff
@@ -107,4 +119,119 @@ public void illegalDefaultPort() {
+        final PropertiesEndpointGroup endpointGroupA = PropertiesEndpointGroup.of(
+                classLoader, file.getName(), "serverA.hosts", 80, true);
+
+        await().atMost(20, TimeUnit.SECONDS).until(() -> endpointGroupA.endpoints().size() == 1);
```
- If there's no good reason to override the default timeout, please remove `atMost()` calls.
- You could use `untilAsserted()` instead of `until()`:

  ```java
  await().untilAsserted(() -> assertThat(endpointGroupA.endpoints()).hasSize(1));
  ```

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroupTest.java:
```diff
@@ -107,4 +119,119 @@ public void illegalDefaultPort() {
+        props.store(printWriter, "");
+        printWriter.close();
+
+        await().atMost(20, TimeUnit.SECONDS).until(() -> endpointGroupA.endpoints().size() == 2);
```
Ditto

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroupTest.java:
```diff
@@ -107,4 +119,119 @@ public void illegalDefaultPort() {
+        final PropertiesEndpointGroup endpointGroupA = PropertiesEndpointGroup.of(
+                classLoader, file.getName(), "serverA.hosts", 80, true);
+
+        await().atMost(20, TimeUnit.SECONDS).until(() -> endpointGroupA.endpoints().size() == 1);
```
Ditto

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroupTest.java:
```diff
@@ -107,4 +119,119 @@ public void illegalDefaultPort() {
+
+        final PropertiesEndpointGroup endpointGroupB = PropertiesEndpointGroup.of(
+                classLoader, file.getName(), "serverB.hosts", 80, true);
+        await().atMost(20, TimeUnit.SECONDS).until(() -> endpointGroupB.endpoints().isEmpty());
```
Ditto

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroupTest.java:
```diff
@@ -107,4 +119,119 @@ public void illegalDefaultPort() {
+        props.store(printWriter, "");
+        printWriter.close();
+
+        await().atMost(20, TimeUnit.SECONDS).until(() -> endpointGroupB.endpoints().size() == 2);
```
Ditto

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroupTest.java:
```diff
@@ -107,4 +119,119 @@ public void illegalDefaultPort() {
+        final StaticEndpointGroup staticEndpointGroup = new StaticEndpointGroup(Endpoint.of("127.0.0.1", 8081));
+        final EndpointGroup endpointGroup = propertiesEndpointGroup.orElse(staticEndpointGroup);
+
+        await().atMost(20, TimeUnit.SECONDS).until(() -> endpointGroup.endpoints().size() == 2);
```
Ditto

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroupTest.java:
```diff
@@ -107,4 +119,119 @@ public void illegalDefaultPort() {
+        props.store(printWriter, "");
+        printWriter.close();
+
+        await().atMost(20, TimeUnit.SECONDS).until(() -> endpointGroup.endpoints().size() == 1);
```
Ditto

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroupTest.java:
```diff
@@ -107,4 +119,119 @@ public void illegalDefaultPort() {
+        props.store(printWriter, "");
+        printWriter.close();
+
+        await().atMost(20, TimeUnit.SECONDS).until(() -> endpointGroup.endpoints().size() == 3);
```
Ditto

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistryTest.java:
```diff
@@ -0,0 +1,111 @@
+        printWriter.print(1);
+        printWriter.close();
+
+        await().atMost(20, TimeUnit.SECONDS).until(() -> val.get() == 1);
```
Ditto

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistryTest.java:
```diff
@@ -0,0 +1,111 @@
+        printWriter.print(2);
+        printWriter.close();
+
+        await().atMost(20, TimeUnit.SECONDS).until(() -> val.get() == 2);
```
Ditto

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroup.java:
```diff
@@ -85,13 +93,42 @@ public static PropertiesEndpointGroup of(ClassLoader classLoader, String resourc
+     */
+    public static PropertiesEndpointGroup of(ClassLoader classLoader, String resourceName,
+                                             String endpointKeyPrefix, int defaultPort,
+                                             boolean reloadable) {
```
Maybe we can just require `Path` instead of `ClassLoader` and `resourceName` to differentiate this from other factory methods that do not reload? What do you think?

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroup.java:
```diff
@@ -141,20 +138,78 @@ public static PropertiesEndpointGroup of(Properties properties, String endpointK
+    private static List<Endpoint> loadEndpoints(Path path, String endpointKeyPrefix, int defaultPort) {
+        final URL resourceUrl;
+        try {
+            resourceUrl = path.toUri().toURL();
```
We could stop converting a `Path` to a `URL` but just use `Files.newInputStream(Path, ...)`.

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroupTest.java:
```diff
@@ -42,7 +42,7 @@
     @BeforeClass
     public static void before() {
-        Awaitility.setDefaultTimeout(20, TimeUnit.SECONDS);
+        Awaitility.setDefaultTimeout(1, TimeUnit.MINUTES);
```
You have to revert this back in `@AfterClass`.

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistryTest.java:
```diff
@@ -37,7 +37,7 @@
     @BeforeClass
     public static void before() {
-        Awaitility.setDefaultTimeout(20, TimeUnit.SECONDS);
+        Awaitility.setDefaultTimeout(1, TimeUnit.MINUTES);
```
You have to revert this back in `@AfterClass`.

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroupTest.java:
```diff
@@ -42,7 +43,12 @@
     @BeforeClass
     public static void before() {
-        Awaitility.setDefaultTimeout(1, TimeUnit.MINUTES);
+        Awaitility.setDefaultTimeout(1, TimeUnit.SECONDS);
```
Is this intended? It seems to me that you don't need to adjust the timeout at all if you are decreasing the timeout.

jrhee17 replied:
whoops! fixed

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,233 @@
+        private RestartableFuture(String namePrefix, Supplier<Runnable> runnableSupplier) {
+            executor = Executors.newSingleThreadExecutor(
+                    new ThreadFactoryBuilder().setDaemon(true)
+                                              .setNameFormat(namePrefix).build());
```
When is this executor's thread terminated? I don't see any `shutdown` or `awaitTermination` calls. Even if the `Runnable` is finished running, the executor thread will stay alive. This causes resource leak in a container environment where an app can be reloaded.

jrhee17 replied:
Ah you're right -- there will be idle threads.
When I tried starting and terminating the `ExecutorService`, it seemed to take too long (5s for each start-stop iteration) so I changed to using `Thread`s which seemed must faster.
Let me know if you have any further concerns

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/RestartableThread.java:
```diff
@@ -0,0 +1,85 @@
+        if (isRunning()) {
+            thread.interrupt();
+            try {
+                thread.join(AWAIT_TERMINATE_MILLIS);
```
Is there a case where this thread is not terminated quickly? How about making sure it's terminated?

```java
while (thread.isAlive()) {
    thread.interrupt();
    try {
        thread.join(1000);
    } catch (InterruptedException e) {
        interrupted = true;
    }
}
thread = null;

if (interrupted) {
    Thread.currentThread().interrupt();
}
```

jrhee17 replied:
good idea! done:)

hyangtack commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,130 @@
+import com.linecorp.armeria.client.endpoint.FileWatcherRunnable.FileWatcherContext;
+
+/**
+ * Wraps a {@code WatchService} and allows paths to be registered.
```
nit: `@code` -> `@link`?

jrhee17 replied:
didn't know about `@link`:( done!

hyangtack commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    /**
+     * Check if future for {@code WatchService} is running.
```
nit: `Returns whether the watching thread is running.`?

hyangtack commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java:
```diff
@@ -0,0 +1,130 @@
+
+    /**
+     * Close watch service, future, and clear registry.
+     * @throws Exception may be thrown if {@code WatchService} is already closed
```
From the javadoc for WatchService.close():
```
     * If this watch service is already closed then invoking this method
     * has no effect.
```
So this statement seems necessary to be checked again?

jrhee17 replied:
good point:)

hyangtack commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRunnable.java:
```diff
@@ -0,0 +1,138 @@
+                    try {
+                        realFilePath = watchedPath.toRealPath();
+                    } catch (IOException e) {
+                        logger.warn("skipping -- unable to get real path for {}", watchedPath);
```
nit: How about passing `e`? `, watchedPath);` -> `, watchedPath, e);`

jrhee17 commented:
Sorry for making big changes frequently after each review:(
**Updates**
- modify `WatchService` logic to not cast `Object` to `Path` incorrectly for `OVERFLOW` type. (this means we won't be able to reload on `OVERFLOW` since we don't know the target directory)
- allow multiple `FileSystem`s to be registered
- move new files to `properties` package
- introduce a "registration key"

jrhee17 commented:
In retrospect, it may be better to leave `PropertiesEndpointGroup` in the `endpoints` package to avoid breaking changes (and since there aren't many relevant files). Let me move those files back to the original package tonight :P

trustin commented:
> Sorry for making big changes frequently after each review:(

No worries. It's always happy to see a pull request evolves in a right direction. :+1:

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java:
```diff
@@ -0,0 +1,228 @@
+     * Key used to register/unregister for file change events.
+     */
+    static final class FileWatchRegisterKey {
+        private final Path filePath;
```
Question: Did you wrap `Path` with `FileWatchRegisterKey` for identity comparison?

jrhee17 replied:
Yes. This was done to allow duplicate path registration. Let me know if I am missing anything (different design pattern, explicitly override to show identity comparison, etc.)

trustin replied:
I see. How about adding some comment about identity comparison in the class Javadoc?

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRunnable.java:
```diff
@@ -0,0 +1,144 @@
+
+                        runCallback(realFilePath);
+                    } else {
+                        if (event.kind().equals(OVERFLOW)) {
```
Could be simplified to `} else if (...) {...} else {...}`

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRunnable.java:
```diff
@@ -0,0 +1,144 @@
+                final boolean reset = key.reset();
+                if (!reset) {
+                    logger.warn("aborting file watch due to unexpected error");
+                    break;
```
Question: Is this an error that prevents us from continue watching? What happens if we do not break the loop here?

jrhee17 replied:
Good point. I wasn't able to create a test case but there was a race condition I was able to reproduce by adding a breakpoint and debugging. 
1. `watchService.take()` returns a `WatchKey`
2. `key.cancel()` is called 
3. `key.reset()` is called and returns false, breaking the loop
This would prevent losing directory watches for other "innocent" paths.

The javadoc imply that `invalid` keys don't take part in further watch events, so I agree it would be better to remove the `break` statement here

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRunnable.java:
```diff
@@ -0,0 +1,144 @@
+            try {
+                ctx.runCallback();
+            } catch (Exception e) {
+                logger.warn("unexpected error from listener: {} ", filePath, e);
```
Could you capitalize the first character of all log messages in this PR? e.g. `unexpected error` -> `Unexpected error`

jrhee17 replied:
done! I didn't apply capitalization to exception messages for now.. but seems like the convention is mixed within the project. would you prefer exception messages be capitalized as well?

trustin replied:
For exception message, please capitalize only when it's a full sentence. No need to capitalize otherwise.

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRunnable.java:
```diff
@@ -0,0 +1,144 @@
+            callback.run();
+        }
+
+        Path getDirPath() {
```
nit: `dirPath()`?

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRunnable.java:
```diff
@@ -0,0 +1,144 @@
+        void cancel() {
+            watchKey.cancel();
+        }
+    }
```
How about overriding `toString()` for easier debugging later?

```java
@Override
public String toString() {
    return MoreObjects.toStringHelper(this).add("dirPath", dirPath)....toString();
```

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java:
```diff
@@ -0,0 +1,228 @@
+         *
+         * @return the path associated with the current key
+         */
+        Path getFilePath() {
```
nit: `filePath()`?

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java:
```diff
@@ -0,0 +1,228 @@
+        Path getFilePath() {
+            return filePath;
+        }
+    }
```
How about overriding `toString()` for easier debugging later?

```java
@Override
public String toString() {
    return MoreObjects.toStringHelper(this).add("filePath", filePath).toString();
```

jrhee17 commented:
**Updates**
- removed `getRealPath` in `FileWatcherRunnable` since we already use the real path when registering  `Path` to `WatchService`
javadoc for `WatchKey.watchable()`:
> Returns the object for which this watch key was created
- I also realized since all events have access to `WatchKey.watchable()`, we can reload files on `OVERFLOW` events
- Decided to pass `context` over a `supplier` for better code readability

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java:
```diff
@@ -74,10 +76,10 @@
-                dirPath = watchRegisterKey.getFilePath().toRealPath().getParent();
+                dirPath = watchRegisterKey.filePath().toRealPath().getParent();
             } catch (IOException e) {
                 throw new IllegalArgumentException("failed to locate file " +
```
For example, this doesn't have to be capitalized because it's not a complete sentence.

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRunnable.java:
```diff
@@ -19,92 +19,80 @@
-                            realFilePath = watchedPath.toRealPath();
-                        } catch (IOException e) {
-                            logger.warn("skipping -- unable to get real path for {}", watchedPath, e);
+                            logger.warn("Ignoring deleted file: {}", filePath);
```
a deleted file

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/FileWatcherRunnableTest.java:
```diff
@@ -19,18 +19,27 @@
+                FileWatchServiceContext.class));
+        when(watchService.take()).then(invocation -> {
+            while (!Thread.currentThread().isInterrupted()) {
+                // do nothing
```
nit: How about adding `Thread.yield()` at least?

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java:
```diff
@@ -0,0 +1,239 @@
+                                            ENTRY_DELETE,
+                                            OVERFLOW);
+            } catch (IOException e) {
+                throw new IllegalArgumentException("failed to watch file " + watchRegisterKey.filePath(), e);
```
a file

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java:
```diff
@@ -0,0 +1,239 @@
+            try {
+                dirPath = watchRegisterKey.filePath().toRealPath().getParent();
+            } catch (IOException e) {
+                throw new IllegalArgumentException("failed to locate file " +
```
a file

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRunnable.java:
```diff
@@ -0,0 +1,141 @@
+                        }
+                        runCallback(filePath);
+                    } else if (event.kind().equals(OVERFLOW)) {
+                        logger.debug("Watch events may have been lost");
```
missing preiod, because it's a full sentence.

trustin commented on core/src/test/java/com/linecorp/armeria/client/endpoint/RestartableThreadTest.java:
```diff
@@ -0,0 +1,48 @@
+        final RestartableThread restartableThread =
+                new RestartableThread(testName.getMethodName(), () -> () -> {
+                    while (!Thread.currentThread().isInterrupted()) {
+                        // do nothing
```
Ditto - `Thread.yield()`?

jrhee17 commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java:
```diff
@@ -0,0 +1,239 @@
+        void register(FileWatchRegisterKey watchRegisterKey, Runnable callback) {
+            final Path dirPath;
+            try {
+                dirPath = watchRegisterKey.filePath().toRealPath().getParent();
```
hmmm.. I think it depends on what the user wants -- the user might want to watch the "real file" disregarding symlinks, or intentionally watch symlink directories that may change. I'm starting to think it might actually be better to delegate this decision to the user and not manipulate the input `path`. (especially since I'm not 100% sure what the use-cases for this class will be)
wdyt?

anuraaga replied:
Yeah, it'd be simple for the user to call `toRealPath` if they wanted that behavior, so not modifying the input SGTM

jrhee17 commented:
**Updates**
- applied comments regarding logging style
- modified to not use real paths
- renamed `FileWatchServiceContext` to `FileSystemWatchContext` for better readability

trustin commented:
@imasahiro Any comments?

imasahiro commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java:
```diff
@@ -0,0 +1,234 @@
+         */
+        void register(FileWatchRegisterKey watchRegisterKey, Runnable callback) {
+            final Path dirPath = watchRegisterKey.filePath().getParent();
+            checkArgument(dirPath != null, "no parent directory for input path: %s",
```
nit. We usualy use `requireNonNull` for null check?

trustin replied:
I think it has to be `IAE` in this case because the user parameter (`filePath`) is not null but its property (`dirPath`) is.

jrhee17 replied:
Originally, I thought it would be better to throw an `IllegalArgumentException` over an NPE to let the user know the input is wrong. If this isn't necessary, I'm fine with sticking to convention:) Any thoughts?

jrhee17 replied:
(It does seem like other input parameters throw NPE though...)
https://github.com/line/armeria/blob/fd6f91f373c03862cbb2a622fd70d8b85c11305c/core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroup.java#L126
`requireNonNull` might be better as it is consistent with other fields validations

jrhee17 replied:
oh wait... let me remove the fix for this. I missed @trustin 's comment:(

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesEndpointGroup.java:
```diff
@@ -181,14 +248,35 @@ private static void validateDefaultPort(int defaultPort) {
+    /**
+     * Unregisters the current {@link PropertiesEndpointGroup} from the {@link FileWatcherRegistry}.
+     * Invoking this method will attempt to release any unused resources used to continuously
+     * watch properties files.
```
Just realized that this refers to `FileWatcherRegistry` which is not a public class, exposing the implementation detail of this class. How about just removing the Javadoc entirely, given that it behaves just like a usual `Closeable`?

minwoox commented on core/src/main/java/com/linecorp/armeria/client/endpoint/RestartableThread.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    /**
+     * Starts thread with runnable if not running yet.
```
nit: a thread?

minwoox commented on core/src/main/java/com/linecorp/armeria/client/endpoint/RestartableThread.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    /**
+     * Whether thread is currently running.
```
a thread?

minwoox commented on core/src/test/java/com/linecorp/armeria/client/endpoint/RestartableThreadTest.java:
```diff
@@ -0,0 +1,48 @@
+public class RestartableThreadTest {
+
+    @Rule
+    public final TestName testName = new TestName();
```
nit: we can use `TestInfo` and convert this test using JUnit 5 if you want.

jrhee17 replied:
👍 although I would have to rebase and I'm a little concerned bc the comments in this PR won't be trackable anymore + it's difficult for reviewers to see change.   If you think rebasing is fine, I can implement in this PR. let me know what you think:)

minwoox replied:
Yeah, it's fine as it is. :-) We were planning to convert all the tests using JUnit 5 eventually, but we can change it later.

jrhee17 commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java:
```diff
@@ -0,0 +1,239 @@
+    }
+
+    private final Map<FileSystem, FileWatchServiceContext> fileSystemWatchServiceMap
+            = new ConcurrentHashMap<>();
```
yes it can since all modifications to this map is in `syncrhonized` calls. 👍

jrhee17 commented on core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java:
```diff
@@ -0,0 +1,239 @@
+                throw new IllegalArgumentException("failed to watch file " + watchRegisterKey.filePath(), e);
+            }
+            currWatchEventMap.put(watchRegisterKey, new FileWatchEvent(watchKey, callback, dirPath));
+            if (!currWatchEventMap.isEmpty()) {
```
you're right, we probably don't need this check.
(Let me know if you also think the method name should be changed to something more appropriate like `startIfPossible`... I think as-is is fine though since a single thread can't start multiple times)

trustin commented:
BOOM! 💥 Thanks again for your hard work and we look forward seeing you again in a new pull request. 😉 👏
